runners/java-fn-execution/src/test/java/org/apache/beam/runners/fnexecution/control/SdkHarnessClientTest.java

```diff
@@ -103,4 +173,149 @@ public void testNewBundleNoDataDoesNotCrash() throws Exception {
+        processor.newBundle(
+            Collections.singletonMap(
+                outputTarget,
+                RemoteOutputReceiver.of(
+                    FullWindowedValueCoder.of(
+                        LengthPrefixCoder.of(StringUtf8Coder.of()), Coder.INSTANCE),
+                    outputs::add)));
+
+    try (CloseableFnDataReceiver<WindowedValue<String>> bundleInputReceiver =
+        (CloseableFnDataReceiver) activeBundle.getInputReceiver()) {
+      bundleInputReceiver.accept(WindowedValue.valueInGlobalWindow("foo"));
+      bundleInputReceiver.accept(WindowedValue.valueInGlobalWindow("bar"));
+      bundleInputReceiver.accept(WindowedValue.valueInGlobalWindow("baz"));
+    }
+    activeBundle.getBundleResponse().get();
+    for (InboundDataClient outputClient : activeBundle.getOutputClients().values()) {
+      outputClient.awaitCompletion();
+    }
+
+    // The bundle can be a simple function of some sort, but needs to be complete.
+    assertThat(
+        outputs,
+        containsInAnyOrder(
+            WindowedValue.valueInGlobalWindow("spam"),
+            WindowedValue.valueInGlobalWindow("ham"),
+            WindowedValue.valueInGlobalWindow("eggs")));
+    executor.shutdownNow();
+  }
+
+  private BeamFnApi.ProcessBundleDescriptor getProcessBundleDescriptor(
+      Endpoints.ApiServiceDescriptor endpoint) throws IOException {
+    Pipeline userPipeline = Pipeline.create();
+    TupleTag<String> outputTag = new TupleTag<>();
+    userPipeline
+        .apply("create", Create.of("foo"))
+        .apply("proc", ParDo.of(new TestFn()).withOutputTags(outputTag, TupleTagList.empty()));
+    RunnerApi.Pipeline userProto = PipelineTranslation.toProto(userPipeline);
+
+    ProcessBundleDescriptor.Builder pbdBuilder =
+        ProcessBundleDescriptor.newBuilder()
+            .setId("my_id")
+            .putAllEnvironments(userProto.getComponents().getEnvironmentsMap())
+            .putAllWindowingStrategies(userProto.getComponents().getWindowingStrategiesMap())
+            .putAllCoders(userProto.getComponents().getCodersMap());
+    RunnerApi.Coder fullValueCoder =
+        CoderTranslation.toProto(WindowedValue.getFullCoder(StringUtf8Coder.of(), Coder.INSTANCE))
+            .getCoder();
+    pbdBuilder.putCoders("wire_coder", fullValueCoder);
+
+    PTransform targetProcessor = userProto.getComponents().getTransformsOrThrow("proc");
+    RemoteGrpcPort port =
+        RemoteGrpcPort.newBuilder()
+            .setApiServiceDescriptor(endpoint)
+            .setCoderId("wire_coder")
+            .build();
+    RemoteGrpcPortRead readNode =
+        RemoteGrpcPortRead.readFromPort(
+            port, Iterables.getOnlyElement(targetProcessor.getInputsMap().values()));
+    RemoteGrpcPortWrite writeNode =
+        RemoteGrpcPortWrite.writeToPort(
+            Iterables.getOnlyElement(targetProcessor.getOutputsMap().values()), port);
+    // TODO: GrpcPortRead to GrpcPortWrite over the Fn API
```

Comment:
Clarify this comment

Reply:
Removed, I think this was me trying to elide the need to actually process anything for a while (this was when the SDK harness didn't take `ParDoPayloads` effectively)